### PR TITLE
ssl: subject alternate name regex validation support

### PR DESCRIPTION
**Pull request by ramaraochavali**
For an explanation of how to fill out the fields, please see the relevant section 
in PULL_REQUESTS.md

*Description*:
Support for regex validation of Subject Alternate names
*Risk Level*: Low
*Testing*: Automated tests added
*Docs Changes*: Docs updated
*Release Notes*: Updated

Fixes https://github.com/envoyproxy/envoy/issues/3705

### CODE REVIEW

File: api/envoy/api/v2/auth/cert.proto [hunk 1]
```diff
@@ -216,6 +216,10 @@ message CertificateValidationContext {
 
   // If specified, Envoy will not reject expired certificates.
   bool allow_expired_certificate = 8;
+
+  // If specified, Envoy will use this regular expression while matching :ref:`Subject Alternative
+  // Names <envoy_api_field_auth.CertificateValidationContext.verify_subject_alt_name>`.
+  string san_validation_regex = 9;
```
Comment (ggreenway) [inline api/envoy/api/v2/auth/cert.proto]: In this file, please put this directly below "verify_subject_alt_name" for readability, and add a comment at the top of the struct for next field:
`// [#comment:next free field: 10]`

Comment (ggreenway) [reply api/envoy/api/v2/auth/cert.proto]: I think you need to clearly define how this interacts with `verify_subject_alt_name`. In the current implementation, if the new setting is supplied, `verify_subject_alt_name` doesn't have any effect. My first thought on how they should interact is that a match on either should be considered valid. Another option is to only allow one of the two settings to be specified and reject config that has both of them. Silently ignoring one in favor of the other seems too confusing for a user.

File: api/envoy/api/v2/auth/cert.proto [hunk 2]
```diff
@@ -216,6 +216,10 @@ message CertificateValidationContext {
 
   // If specified, Envoy will not reject expired certificates.
   bool allow_expired_certificate = 8;
+
+  // If specified, Envoy will use this regular expression while matching :ref:`Subject Alternative
```
Comment (ggreenway) [inline api/envoy/api/v2/auth/cert.proto]: This setting should have the same constraint and warning as `verify_subject_alt_name`, so I think those parts of the documentation should be copied here.

**Comment by PiotrSikora** [comment]
@ggreenway I disagree, compare this PR with #1298.

Also, there is a question whether or not we need to use a separate field for this (I think not).

File: source/common/ssl/context_impl.cc [hunk 3]
```diff
@@ -351,8 +358,14 @@ bool ContextImpl::verifySubjectAltName(X509* cert,
       ASN1_STRING* str = san->d.uniformResourceIdentifier;
       const char* uri = reinterpret_cast<const char*>(ASN1_STRING_data(str));
       for (auto& config_san : subject_alt_names) {
-        if (config_san.compare(uri) == 0) {
-          return true;
+        if (is_san_regex_validation) {
+          if (std::regex_match(config_san, regex_pattern)) {
```
Comment (ggreenway) [inline source/common/ssl/context_impl.cc]: This is matching the wrong field: the regex should be matched against `uri`, not against the configuration-supplied `config_san`.

**Comment by ggreenway** [comment]
@PiotrSikora Ok, that makes sense.

@ramaraochavali Let's hold off on the comments I made about the implementation for now, and figure out the right feature and how to configure it first.

**Comment by ramaraochavali** [comment]
@ggreenway Ok. Sure no problem. I will close this PR for now and let us discuss on the issue how this should be configured. Once that is decided, I will open a fresh PR with the correct implementation.

**Pull request closed without merging**